File: favourite-genres.cpp
// Time Complexity: O(kmn)
// Space Complexity: O(mn)
// k = #users
// m = #genres
// n = average #songs in each genre
unordered_map<string, vector<string>> favouriteGenres(unordered_map<string, vector<string>>& userSongs, unordered_map<string, vector<string>>& songGenres) {
    unordered_map<string, vector<string>> result;
    unordered_map<string, string> songToGenre; // space : O(mn)
    for(auto songGenre : songGenres) { // time : O(mn)
        string genre = songGenre.first;
        vector<string> songs = songGenre.second;
        for(auto song : songs) {
            songToGenre[song] = genre;
        }
    }
    // time : O(kmn)
    for(auto userSong : userSongs) { // O(k)
        string user = userSong.first; // name of the user
        // list of particular user's favourite songs
        vector<string> songs = userSong.second;
        // count of songs(fav. songs of user) belonging to each genre
        // key : genre; value : count of songs
        unordered_map<string, int> count; // space : O(m)
        int maxCnt = 0; // maximum count of songs among all the genres
        // In worst case, user can like all mn songs
        for(auto song : songs) { // O(mn)
            count[songToGenre[song]]++;
            maxCnt = max(maxCnt, count[songToGenre[song]]);
        }
        
        for(auto i : count) { // O(m)
            string genre = i.first; // name of the genre
            int nSongs = i.second; // number of songs belonging to a genre
            if(maxCnt == nSongs) {
                result[user].push_back(genre);
            }
        }
    }
    return result;
}

int main() {
    unordered_map<string, vector<string>> userSongs;
    userSongs["David"] = {"song1", "song2", "song3", "song4", "song8"};
    userSongs["Emma"] = {"song5", "song6", "song7"};
    
    unordered_map<string, vector<string>> songGenres;
    songGenres["Rock"] = {"song1", "song3"};
    songGenres["Dubstep"] = {"song7"};
    songGenres["Techno"] = {"song2", "song4"};
    songGenres["Pop"] = {"song5", "song6"};
    songGenres["Jazz"] = {"song8", "song9"};
    
    unordered_map<string, vector<string>> result;
    result = favouriteGenres(userSongs, songGenres);
    for(auto i : result) {
        cout << i.first << " : ";
        for(auto genre : i.second) {
            cout << genre << ", ";
        }
        cout << endl;
    }
}

File: Favourite Genres.java
public class Main {
    
    private static HashMap<String, List<String>> findFavGenre(HashMap<String, List<String>> userToSongs, HashMap<String, List<String>> genreToSongs)
    {
        HashMap<String, String> songsToGenre = new HashMap();
        HashMap<String, List<String>> result = new HashMap();
        
        for(String genre : genreToSongs.keySet())
        {
            List<String> songs = genreToSongs.get(genre);
            
            for(String song : songs) songsToGenre.put(song, genre);
        }
        
        for(String user : userToSongs.keySet())
        {
            List<String> songs = userToSongs.get(user);
            
            HashMap<String, Integer> map = new HashMap();
            int maximum = 0;
            
            for(String song : songs)
            {
                String genre = songsToGenre.get(song);
                
                map.put(genre, map.getOrDefault(genre, 0) + 1);
                
                maximum = Math.max(maximum, map.get(genre));
            }
            
            for(String genre : map.keySet())
            {
                if(map.get(genre) == maximum)
                {
                    if(!result.containsKey(user)) result.put(user, new ArrayList());
                    
                    result.get(user).add(genre);
                }
            }
        }
        
        return result;
    }
    public static void main(String[] args) {
        
        HashMap<String, List<String>> userToSongs = new HashMap();
        HashMap<String, List<String>> genreToSongs = new HashMap();
        
        userToSongs.put("Malcolm", new ArrayList(Arrays.asList("Song1", "Song2", "Song3", "Song4")));
        userToSongs.put("Daisy", new ArrayList(Arrays.asList("Song5", "Song6", "Song7")));
        userToSongs.put("Marshall", new ArrayList(Arrays.asList("Song8", "Song9", "Song10")));
        
        genreToSongs.put("Rock", new ArrayList(Arrays.asList("Song7", "Song5", "Song6", "Song8")));
        genreToSongs.put("Pop", new ArrayList(Arrays.asList("Song2", "Song1", "Song9")));
        genreToSongs.put("Jazz", new ArrayList(Arrays.asList("Song3", "Song4", "Song10")));
        
        HashMap<String, List<String>> userToGenre = findFavGenre(userToSongs, genreToSongs);
        
        System.out.println(userToGenre);
    }
}